src/Symfony/Component/Validator/Test/CompoundConstraintTestCase.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Test;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Validator\Constraints\Compound;
+use Symfony\Component\Validator\Constraints\GroupSequence;
+use Symfony\Component\Validator\ConstraintViolationInterface;
+use Symfony\Component\Validator\ConstraintViolationListInterface;
+use Symfony\Component\Validator\Validation;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+/**
+ * A test case to ease testing Compound Constraint.
+ *
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ */
+abstract class CompoundConstraintTestCase extends TestCase
+{
+    protected ValidatorInterface $validator;
+
+    protected ?ConstraintViolationListInterface $violationList = null;
+
+    protected function setUp(): void
+    {
+        $this->validator = Validation::createValidator();
+    }
+
+    protected function validateValue(mixed $value, array|string|GroupSequence $groups = null): void
+    {
+        $this->violationList = $this->validator->validate($value, $this->createCompound(), $groups);
+    }
+
+    protected function assertViolationIsRaisedByCompound(string $code): void
+    {
+        if (null === $this->violationList) {
+            throw new \LogicException('Value must be validated with the "validateValue()" method before asserting on compound violations.');
+        }
+
+        $this->assertTrue($this->isViolationRaised($code), sprintf('Expected violation with code "%s" to be raised.', $code));
+    }
+
+    protected function assertViolationIsNotRaisedByCompound(string $code): void
```

Comment:
I'm not sure using codes is the best API here. the violation code is optional.

this should rather use an API similar to the ConstraintValidatorTestCase to assert violations

Reply:
Refactored completely the internals to be able to assert on `Constraint` objects directly, as shown in the test